src/Algebra/GCD.agda

```diff
@@ -0,0 +1,43 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Definition of GCD.
+------------------------------------------------------------------------
+
+-- You're unlikely to want to use this module directly. Instead you
+-- probably want to be importing the appropriate module from
+-- `Algebra.Properties.CancellativeCommutativeSemiring`.
+
+{-# OPTIONS --without-K --safe #-}
+
+open import Algebra.Core using (Op₂)
+open import Data.Product using (proj₁; proj₂)
+open import Level using (_⊔_)
+open import Relation.Binary using (Rel)
+
+module Algebra.GCD {a ℓ} {A : Set a} (_≈_ : Rel A ℓ) (_*_ : Op₂ A) where
+
+open import Algebra.Divisibility _≈_ _*_
+
+------------------------------------------------------------------------------
+-- Definitions
+
+record GCD (arg1 arg2 : A) :  Set (a ⊔ ℓ) where      -- a result of gcd
```

Comment:
Would it make sense to have an `IsGCD x y z` record that had the GCD value in the type as well? Then you could define `GCD` in terms of `IsGCD`?

Reply:
I was glad that you did not not recall of such a thing. And now you do.
Do you suggest 
```
record IsGCD (x y z : A) : Set _ where                                            
  field                                                                         
    commonDivisor : z ∣ x × z ∣ y 
    greatest      : ∀ {u} → u ∣ x → u ∣ y → u ∣ z
```
(similar to Nat.GCD in `master`) ?
And then, how do you define the GCD type and the function
```gcd : (x y : A) → GCD x y
```
in the GCDSemiring record?  IsGCD and GCD - to much of them, complication ...

Traditionally, there is only GCD with two parameters, and  gcd x y  returns  GCD x y.
Classically, it returns a value in the type A,
but in constructive algebra, it should return a structure which also contains witnesses.

Can you, please, remain GCD as it is: 
```
record GCD (x y : A) :  Set (a ⊔ ℓ) where  
  constructor gcdᶜ                                                                
  field                                                                           
    value    : A          
    divides₁ : value ∣ x  
    divides₂ : value ∣ y     
    greatest : ∀ {z} → z ∣ x → z ∣ y → z ∣ value
  ...
```
?

Reply:
Yes I suggest doing it the same way we do it in `Data.Nat.GCD`.

>  And then, how do you define the GCD type and the function, gcd : (x y : A) → GCD x y

I would either go
```agda
  gcd : A → A → A
  gcd-isGCD : IsGCD x y (gcd x y)
```
or
```agda
  gcd : ∃ λ g → IsGCD x y g
```
I prefer the first because it allows you to access the raw `gcd` function without having to project out of the pair.

The advantage of having the value as a parameter to the record is that it's much easier to write types for the proofs. For example consider the following result:
```agda
GCD-* : ∀ {m n d c} → IsGCD (m * c) (n * c) (d * c) → IsGCD m n d
```
Using your current definition, you would have to write something like:
```agda
GCD-* : ∀ {m n c} (g1: GCD (m * c) (n * c)) (g2 : GCD m n d) → GCD.value g1 ≡ c * GCD.value g2
```
naming the input records, and extracting them out of the bundle, and resulting in an equality rather than a new `IsGCD` instance.